File: sdk-java/src/main/java/io/superdurable/dex/StopFlowOptions.java
/*
 * Copyright (c) 2026 Super Durable, Inc.
 *
 * Licensed under the Super Durable Source License 1.0.
 * You may not use this file except in compliance with the License.
 * See the LICENSE file in the repository root.
 *
 * SPDX-License-Identifier: LicenseRef-Super-Durable-1.0
 */

package io.superdurable.dex;

/**
 * Configures how {@link Client#stopFlow} ends a running Flow.
 *
 * <p>The default constructor requests cancellation without a reason. Use the explicit constructor
 * to terminate or fail a Flow and to attach a human-readable reason.
 *
 * <pre>{@code
 * client.stopFlow(
 *         "order-123",
 *         new StopFlowOptions(StopType.TERMINATE, "operator request"));
 * }</pre>
 */
public final class StopFlowOptions {
    private final StopType type;
    private final String reason;

    /** Creates cancellation options without an explicit reason. */
    public StopFlowOptions() {
        this(StopType.CANCEL, null);
    }

    /**
     * Creates stop options with an explicit mode and reason.
     *
     * @param type the requested terminal behavior
     * @param reason the user-visible reason, or {@code null}
     */
    public StopFlowOptions(final StopType type, final String reason) {
        this.type = type;
        this.reason = reason;
    }

    StopType getType() {
        return type;
    }

    String getReason() {
        return reason;
    }
}

File: sdk-rust/crates/dex-sdk/src/stop_flow_options.rs
// Copyright (c) 2026 Super Durable, Inc.
//
// Licensed under the Super Durable Source License 1.0.
// You may not use this file except in compliance with the License.
// See the LICENSE file in the repository root.
//
// SPDX-License-Identifier: LicenseRef-Super-Durable-1.0

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub(crate) enum StopType {
    Cancel,
    Terminate,
    Fail,
}

#[derive(Clone, Debug, Eq, PartialEq)]
/// Selects how [`crate::Client::stop_flow`] ends an active Flow.
///
/// Construct a value with [`Self::cancel`], [`Self::terminate`], or [`Self::fail`], then optionally
/// attach an operator-facing reason.
pub struct StopFlowOptions {
    pub(crate) stop_type: StopType,
    pub(crate) reason: Option<String>,
}

impl StopFlowOptions {
    fn new(stop_type: StopType) -> Self {
        Self {
            stop_type,
            reason: None,
        }
    }

    /// Requests cooperative cancellation so application cleanup can run.
    pub fn cancel() -> Self {
        Self::new(StopType::Cancel)
    }

    /// Terminates the Flow immediately without cooperative cleanup.
    pub fn terminate() -> Self {
        Self::new(StopType::Terminate)
    }

    /// Marks the Flow failed with the reason set by [`Self::reason`].
    pub fn fail() -> Self {
        Self::new(StopType::Fail)
    }

    /// Attaches a reason recorded with the stop operation.
    pub fn reason(mut self, value: impl Into<String>) -> Self {
        self.reason = Some(value.into());
        self
    }
}
